### PR TITLE
Feature/msg builders

### PR DESCRIPTION
**Pull request by bobozaur**
This PR introduces message builders through the `typed-builder` crate. The builder are based on the typestate pattern, but the crate is (at least advertised) to be fully compatible/replaceable by the `derive-builder` trait, which is using runtime construction of the type and allows for more flexibility (like passing builders around).

The `typestate` vs `runtime construction` decision will essentially depend on whether consumers want/need to build messages themselves and how convoluted these processes usually get. We don't really have to decide now, but it's a good thing to consider in the future.

For consistency reasons even some single fielded protocol specific messages contents implement builders, so that it is convenient, uniform and easy to know how one would go about composing a message.

There are some inner types which also implement the builders, yet for some types it is indeed quite the overkill (like `CredentialPreview`). I might revisit such types and just implement plain old `new()` constructors.

Another change that this PR brings is a uniformization of the `Invitation` message from the `Connection` protocol, so that the same decorators are used regardless of invitation type. The rationale is that it's much easier to reason with the invitation this way and since the protocol is old and won't be suffering changes it should be fine to simply deal with it like this.

Also, this addresses #949 .

### CODE REVIEW

**Review by Patrik-Stas — DISMISSED** [review]
This is great! Looks a lot cleaner, much easier to read. Very nice 👍

**Review by mirgee — DISMISSED** [review]
Love it! Makes message construction a breeze - so much easier on the eyes, the programmer is easily guided by the code completion now and I like that it is guaranteed that all mandatory fields are set at compile time 👍

File: aries_vcx/src/protocols/mediated_connection/invitee/state_machine.rs [hunk 1]
```diff
@@ -223,34 +223,39 @@ impl SmConnectionInvitee {
...
+                let content = RequestContent::builder()
+                    .label(self.source_id.to_string())
+                    .connection(con_data)
+                    .build();
 
-                let mut decorators = RequestDecorators::default();
-                let mut timing = Timing::default();
-                timing.out_time = Some(Utc::now());
-                decorators.timing = Some(timing);
+                let decorators = RequestDecorators::builder().timing(Timing::builder().out_time(Utc::now()).build());
```
[the remark this answers is not in the document]
Comment (mirgee) [reply aries_vcx/src/protocols/mediated_connection/invitee/state_machine.rs:231]: IMO, not unless maximum line width is exceeded. Just let rustfmt take care of formatting. We may consider lowering our workspace-wide max line width from 120 to the default 100, though.

[the remark this answers is not in the document]
Comment (bobozaur) [reply aries_vcx/src/protocols/mediated_connection/invitee/state_machine.rs:231]: Not intentional, really just rustfmt doing its thing.

File: aries_vcx/src/protocols/connection/invitee/mod.rs [hunk 2]
```diff
@@ -120,21 +120,33 @@ impl InviteeConnection<Invited> {
...
+                    let thread = Thread::builder()
+                        .thid(id.clone())
+                        .pthid(self.state.thread_id().to_owned())
+                        .build();
+
+                    (id.clone(), thread)
+                }
+                InvitationContent::Pairwise(_) | InvitationContent::PairwiseDID(_) => {
+                    let thread = Thread::builder().thid(self.state.thread_id().to_owned()).build();
+                    (self.state.thread_id().to_owned(), thread)
```
[the remark this answers is not in the document]
Comment (bobozaur) [reply aries_vcx/src/protocols/connection/invitee/mod.rs:138]: Sounds good. On it.

[the remark this answers is not in the document]
Comment (bobozaur) [reply aries_vcx/src/protocols/connection/invitee/mod.rs]: Done.

**Review by Patrik-Stas — APPROVED** [review]

**Review by mirgee — APPROVED** [review]

**Pull request merged**